File: src/main/java/me/centralhardware/znatoki/telegram/statistic/service/ClientService.java
package me.centralhardware.znatoki.telegram.statistic.service;

import jakarta.persistence.EntityManager;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import me.centralhardware.znatoki.telegram.statistic.entity.Client;
import me.centralhardware.znatoki.telegram.statistic.repository.ClientRepository;
import org.hibernate.search.mapper.orm.Search;
import org.hibernate.search.mapper.orm.session.SearchSession;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

import static java.util.function.Predicate.not;

@Service
@Slf4j
@RequiredArgsConstructor
public class ClientService {


    private final ClientRepository repository;
    private final EntityManager entityManager;

    public String getFioById(Integer id){
        return findById(id)
                .map(Client::getFio)
                .orElse("");
    }

    public List<Client> getAll(){
        return repository.findAll()
                .stream()
                .filter(result -> !result.isDeleted())
                .toList();
    }

    public Client save(Client client){
        CompletableFuture.runAsync(this::updateIndex);
        return repository.save(client);
    }

    @SuppressWarnings("unchecked")
    @Transactional
    public List<Client> search(String text) {
        var query  = text.toLowerCase();

        SearchSession session = Search.session(entityManager);

        return session.search(Client.class)
                .where(it -> it.match()
                        .field("name")
                        .field("secondName")
                        .field("lastName")
                        .matching(query)
                        .fuzzy())
                .fetch(10)
                .hits();
    }

    @Scheduled(fixedRate = 60, initialDelay = 1, timeUnit = TimeUnit.MINUTES)
    @Transactional
    public void updateIndex() {
        try {
            log.info("update lucene index");
            Search.session(entityManager).massIndexer().startAndWait();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public Optional<Client> findById(Integer id){
        return repository.findById(id)
                .filter(not(Client::isDeleted));
    }

    public Client findByFioAndId(String fio){
        return repository.findByFioAndId(fio);
    }

    public boolean checkExistenceByFio(String name, String secondName, String lastName){
        return repository.findAllByNameAndSecondNameAndLastName(name, secondName, lastName)
                .stream()
                .anyMatch(it -> !it.isDeleted());

    }


}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/service/ClientService.kt
package me.centralhardware.znatoki.telegram.statistic.service

import java.io.IOException
import kotlin.concurrent.fixedRateTimer
import me.centralhardware.znatoki.telegram.statistic.entity.Student
import me.centralhardware.znatoki.telegram.statistic.mapper.StudentMapper
import org.apache.lucene.analysis.standard.StandardAnalyzer
import org.apache.lucene.document.Document
import org.apache.lucene.document.Field
import org.apache.lucene.document.TextField
import org.apache.lucene.index.*
import org.apache.lucene.search.*
import org.apache.lucene.store.ByteBuffersDirectory

object ClientService {

    private var directory: ByteBuffersDirectory? = null

    fun init() {
        fixedRateTimer(name = "updateLucene", period = 60000L) {
            val index = ByteBuffersDirectory()
            val analyzer = StandardAnalyzer()
            val indexWriterConfig = IndexWriterConfig(analyzer)
            val writer = IndexWriter(index, indexWriterConfig)

            StudentMapper.findAll().forEach { client ->
                try {
                    val document = Document()
                    document.add(TextField("name", client.name.lowercase(), Field.Store.YES))
                    document.add(
                        TextField("lastName", client.lastName.lowercase(), Field.Store.YES)
                    )
                    document.add(
                        TextField("secondName", client.secondName.lowercase(), Field.Store.YES)
                    )
                    document.add(TextField("id", client.id.toString(), Field.Store.YES))
                    writer.addDocument(document)
                } catch (e: IOException) {
                    throw java.lang.RuntimeException(e)
                }
            }
            writer.close()
            directory = index
        }
    }

    fun search(fio: String): List<Student> {
        // Создаем поисковые запросы для каждого слова по каждому полю (имя, фамилия и отчество)
        val queries =
            fio.split(" ").flatMap { word ->
                listOf(
                    FuzzyQuery(Term("name", word), 2),
                    FuzzyQuery(Term("secondName", word), 2),
                    FuzzyQuery(Term("lastName", word), 2),
                )
            }

        // Конструируем BooleanQuery, объединяя все запросы
        val combinedQuery = BooleanQuery.Builder()
        queries.forEach { query -> combinedQuery.add(query, BooleanClause.Occur.SHOULD) }

        val indexReader: IndexReader = DirectoryReader.open(directory)
        val searcher = IndexSearcher(indexReader)
        val topDocs = searcher.search(combinedQuery.build(), 10)
        return topDocs.scoreDocs
            .sortedBy { it.score }
            .reversed()
            .map { searcher.storedFields().document(it.doc) }
            .mapNotNull { it: Document -> StudentMapper.findById(it["id"].toInt()) }
            .toList()
    }
}
